File: saunity/src/main/java/tv/superawesome/plugins/publisher/unity/SAUnityInterstitialAd.java
/*
 * @Copyright: SADefaults Trading Limited 2017
 * @Author: Gabriel Coman (gabriel.coman@superawesome.tv)
 */
package tv.superawesome.plugins.publisher.unity;

import android.content.Context;

import org.json.JSONException;
import org.json.JSONObject;

import tv.superawesome.lib.sasession.defines.SAConfiguration;
import tv.superawesome.plugins.publisher.unity.util.SAJsonUtil;
import tv.superawesome.sdk.publisher.SAEvent;
import tv.superawesome.sdk.publisher.SAInterface;
import tv.superawesome.sdk.publisher.SAInterstitialAd;
import tv.superawesome.sdk.publisher.SAOrientation;
import tv.superawesome.sdk.publisher.state.CloseButtonState;

/**
 * Class that holds a number of static methods used to communicate with Unity
 */
public class SAUnityInterstitialAd {

    private static final String unityName = "SAInterstitialAd";

    /**
     * Method that creates a new Interstitial Ad (from Unity)
     */
    public static void SuperAwesomeUnitySAInterstitialAdCreate(Context context) {
        SAInterstitialAd.setListener((SAInterface) (placementId, event) -> {
            switch (event) {
                case adLoaded:
                    SAUnityCallback.sendAdCallback(unityName, placementId, SAEvent.adLoaded.toString());
                    break;
                case adEmpty:
                    SAUnityCallback.sendAdCallback(unityName, placementId, SAEvent.adEmpty.toString());
                    break;
                case adFailedToLoad:
                    SAUnityCallback.sendAdCallback(unityName, placementId, SAEvent.adFailedToLoad.toString());
                    break;
                case adAlreadyLoaded:
                    SAUnityCallback.sendAdCallback(unityName, placementId, SAEvent.adAlreadyLoaded.toString());
                    break;
                case adShown:
                    SAUnityCallback.sendAdCallback(unityName, placementId, SAEvent.adShown.toString());
                    break;
                case adFailedToShow:
                    SAUnityCallback.sendAdCallback(unityName, placementId, SAEvent.adFailedToShow.toString());
                    break;
                case adClicked:
                    SAUnityCallback.sendAdCallback(unityName, placementId, SAEvent.adClicked.toString());
                    break;
                case adEnded:
                    SAUnityCallback.sendAdCallback(unityName, placementId, SAEvent.adEnded.toString());
                    break;
                case adClosed:
                    SAUnityCallback.sendAdCallback(unityName, placementId, SAEvent.adClosed.toString());
                    break;
            }
        });
    }

    /**
     * Method that loads a new Interstitial AD (from Unity)
     */
    public static void SuperAwesomeUnitySAInterstitialAdLoad(Context context,
                                                             int placementId,
                                                             int configuration,
                                                             boolean test,
                                                             String openRtbPartnerId,
                                                             String encodedOptions) {
        SAInterstitialAd.setTestMode(test);
        SAInterstitialAd.setConfiguration(SAConfiguration.fromValue(configuration));

        if (encodedOptions != null && !encodedOptions.isEmpty()) {
            try {
                SAInterstitialAd.load(
                        placementId,
                        context,
                        openRtbPartnerId,
                        SAJsonUtil.JSONtoMap(new JSONObject(encodedOptions))
                );
            } catch (JSONException e) {
                e.printStackTrace();
                // Fallback to loading without options
                SAInterstitialAd.load(placementId, context, openRtbPartnerId);
            }
        } else {
            SAInterstitialAd.load(placementId, context, openRtbPartnerId);
        }
    }

    /**
     * Method that loads a new Interstitial AD (from Unity)
     */
    public static void SuperAwesomeUnitySAInterstitialAdLoad(Context context,
                                                             int placementId,
                                                             int configuration,
                                                             boolean test,
                                                             String encodedOptions) {
        SuperAwesomeUnitySAInterstitialAdLoad(
                context,
                placementId,
                configuration,
                test,
                encodedOptions,
                null
        );
    }

    /**
     * Method that checks to see if an ad is available for an interstitial ad (from Unity)
     */
    public static boolean SuperAwesomeUnitySAInterstitialAdHasAdAvailable(Context context, int placementId) {
        return SAInterstitialAd.hasAdAvailable(placementId);
    }

    /**
     * Method that plays a new Interstitial Ad (from Unity)
     */
    public static void SuperAwesomeUnitySAInterstitialAdPlay(Context context,
                                                             int placementId,
                                                             boolean isParentalGateEnabled,
                                                             boolean isBumperPageEnabled,
                                                             int orientation,
                                                             boolean isBackButtonEnabled) {
        SAInterstitialAd.setParentalGate(isParentalGateEnabled);
        SAInterstitialAd.setBumperPage(isBumperPageEnabled);
        SAInterstitialAd.setOrientation(SAOrientation.fromValue(orientation));
        SAInterstitialAd.setBackButton(isBackButtonEnabled);
        SAInterstitialAd.play(placementId, context);
    }

    /**
     * Method that apply settings (from Unity)
     */
    public static void SuperAwesomeUnitySAInterstitialAdApplySettings(
            boolean isParentalGateEnabled,
            boolean isBumperPageEnabled,
            int orientation,
            boolean isBackButtonEnabled,
            boolean testModeEnabled,
            int closeButtonState) {
        SAInterstitialAd.setParentalGate(isParentalGateEnabled);
        SAInterstitialAd.setBumperPage(isBumperPageEnabled);
        SAInterstitialAd.setBackButton(isBackButtonEnabled);
        SAInterstitialAd.setOrientation(SAOrientation.fromValue(orientation));
        SAInterstitialAd.setTestMode(testModeEnabled);
        setCloseButtonState(closeButtonState);
    }

    private static void setCloseButtonState(int closeButtonState) {
        switch (CloseButtonState.fromInt(closeButtonState)) {
            case Hidden:
                // Do nothing as Interstitial does not support hidden close button
                break;
            case VisibleImmediately:
                SAInterstitialAd.enableCloseButtonNoDelay();
                break;
            case VisibleWithDelay:
                SAInterstitialAd.enableCloseButton();
                break;
        }
    }
}

File: superawesome-unity/src/main/java/tv/superawesome/plugins/publisher/unity/SAUnityInterstitialAd.kt
/**
 * @Copyright:   SADefaults Trading Limited 2017
 * @Author:      Gabriel Coman (gabriel.coman@superawesome.tv)
 */
package tv.superawesome.plugins.publisher.unity

import android.content.Context
import tv.superawesome.sdk.publisher.common.models.Orientation
import tv.superawesome.sdk.publisher.common.models.SAEvent
import tv.superawesome.sdk.publisher.common.models.SAInterface
import tv.superawesome.sdk.publisher.ui.interstitial.SAInterstitialAd.hasAdAvailable
import tv.superawesome.sdk.publisher.ui.interstitial.SAInterstitialAd.load
import tv.superawesome.sdk.publisher.ui.interstitial.SAInterstitialAd.play
import tv.superawesome.sdk.publisher.ui.interstitial.SAInterstitialAd.setBackButton
import tv.superawesome.sdk.publisher.ui.interstitial.SAInterstitialAd.setBumperPage
import tv.superawesome.sdk.publisher.ui.interstitial.SAInterstitialAd.setListener
import tv.superawesome.sdk.publisher.ui.interstitial.SAInterstitialAd.setOrientation
import tv.superawesome.sdk.publisher.ui.interstitial.SAInterstitialAd.setParentalGate
import tv.superawesome.sdk.publisher.ui.interstitial.SAInterstitialAd.setTestMode

/**
 * Class that holds a number of static methods used to communicate with Unity
 */
object SAUnityInterstitialAd {
    private const val unityName = "SAInterstitialAd"

    /**
     * Method that creates a new Interstitial Ad (from Unity)
     */
    fun SuperAwesomeUnitySAInterstitialAdCreate(context: Context?) {
        setListener(object : SAInterface {
            override fun onEvent(placementId: Int, event: SAEvent) {
                when (event) {
                    SAEvent.AdLoaded -> SAUnityCallback.sendAdCallback(unityName, placementId, SAEvent.AdLoaded.toString())
                    SAEvent.AdEmpty -> SAUnityCallback.sendAdCallback(unityName, placementId, SAEvent.AdEmpty.toString())
                    SAEvent.AdFailedToLoad -> SAUnityCallback.sendAdCallback(unityName, placementId, SAEvent.AdFailedToLoad.toString())
                    SAEvent.AdAlreadyLoaded -> SAUnityCallback.sendAdCallback(unityName, placementId, SAEvent.AdAlreadyLoaded.toString())
                    SAEvent.AdShown -> SAUnityCallback.sendAdCallback(unityName, placementId, SAEvent.AdShown.toString())
                    SAEvent.AdFailedToShow -> SAUnityCallback.sendAdCallback(unityName, placementId, SAEvent.AdFailedToShow.toString())
                    SAEvent.AdClicked -> SAUnityCallback.sendAdCallback(unityName, placementId, SAEvent.AdClicked.toString())
                    SAEvent.AdEnded -> SAUnityCallback.sendAdCallback(unityName, placementId, SAEvent.AdEnded.toString())
                    SAEvent.AdClosed -> SAUnityCallback.sendAdCallback(unityName, placementId, SAEvent.AdClosed.toString())
                }
            }
        })
    }

    /**
     * Method that loads a new Interstitial AD (from Unity)
     */
    fun SuperAwesomeUnitySAInterstitialAdLoad(context: Context?, placementId: Int, configuration: Int, test: Boolean) {
        setTestMode(test)
        load(placementId, context!!)
    }

    /**
     * Method that checks to see if an ad is available for an interstitial ad (from Unity)
     */
    fun SuperAwesomeUnitySAInterstitialAdHasAdAvailable(context: Context?, placementId: Int): Boolean {
        return hasAdAvailable(placementId)
    }

    /**
     * Method that plays a new Interstitial Ad (from Unity)
     */
    fun SuperAwesomeUnitySAInterstitialAdPlay(context: Context?, placementId: Int, isParentalGateEnabled: Boolean, isBumperPageEnabled: Boolean, orientation: Int, isBackButtonEnabled: Boolean) {
        setParentalGate(isParentalGateEnabled)
        setBumperPage(isBumperPageEnabled)
        Orientation.fromValue(orientation)?.let { setOrientation(it) }
        setBackButton(isBackButtonEnabled)
        play(placementId, context!!)
    }
}